global-script.lua

```diff
@@ -3312,3 +3467,297 @@ function tCompare(t1,t2)
+    swapPlayerAreaObjects(a, b)
+    swapPlayerAreaColors(a, b)
+    printToAll(a .. " swapped places with " .. b .. ".", Color[a])
+end
+
+function swapPlayerPresenceColors(fromColor, toColor)
+    if fromColor == toColor then return end 
+    local function initData(color, ix, oppositeColor)
+        bag = getObjectFromGUID(PlayerBags[color])
+        return {
+            color = color,
+            ix = ix,
+            bag = bag,
+            qty = bag.getQuantity(),
+            tints = {},
+            objects = {},
+            pattern = color .. "'s (.*)",
+            bagContents = {},
+            defendTint = defendBags[oppositeColor].getColorTint(),
+            oppositeColor = oppositeColor
+        }
+    end
+    local colors = {
+        from = initData(fromColor, 1, toColor),
+        to = initData(toColor, 2, fromColor)
+    }
+
+    -- If both bags are full, there's not a lot of work to do.
+    -- Unfortunately, we still need to loop through other things because of defend tokens that aren't in bags.
+    local fastSwap = (colors.from.qty == 20 and colors.to.qty == 20)
+    -- Just bail out fast.
+
+    readyTokens[fromColor], readyTokens[toColor] = readyTokens[toColor], readyTokens[fromColor]
+    selectedColors[fromColor], selectedColors[toColor] = selectedColors[toColor], selectedColors[fromColor]
+
+    if not fastSwap then
+        -- Remove any items still in the bags
+        -- NOTE: TTS's documentation suggests we may need to wait a physics frame after doing this,
+        -- but this seems to work fine without doing that.  If something goes awry with this code in the future,
+        -- you might try re-adding the delay.
+        for color,data in pairs(colors) do
+            for i = 1,data.qty do 
+                local obj = data.bag.takeObject({
+                    sound=false,
+                    position={x=data.ix*2, z=200, y=i*2}    -- Chosen to be out-of-the-way and to prevent items from stacking.
+                })
+                table.insert(data.bagContents, obj)
+            end
+        end
+    end
+
+    -- Pass 1: Iterate over all objects looking for "<color>'s X".
+    -- Make a note of what we find and what tint it is.  Handle Defence tokens in this pass.
+    local match = string.match  -- Performance
+    local name, suffix
+    for _,obj in pairs(getAllObjects()) do
+        name = obj.getName()
+        if name then
+            for _,data in pairs(colors) do 
+                suffix = match(name, data.pattern)
+                if suffix then
+                    if suffix == 'Defence' then  -- Defense tokens are special.
+                        -- local state = obj.getStateId()
+                        -- local attrs = {position = obj.getPosition(), rotation = obj.getRotation(), smooth = false}
+                        -- destroyObject(obj)
+                        -- data.defendBag.takeObject(attrs)
+                        obj.setColorTint(data.defendTint)
```

Comment:
this is only changing the current state of the defend token right? so if i swap to another state i'll see my original color again?

Reply:
State of the Defend tokens is their defend value (1-20 or Infinite).  It has nothing to do with color.

Reply:
giving this further thought, i think changing only the current state is probably acceptable. these are ephemeral tokens that once placed don't really have their state change, and are cleaned up every turn

Reply:
Ohh.  I _just_ realized what you're actually asking here.  It does indeed only change on state.  If we swap to the approach that actually destroys/recreates the tokens that'll be fixed.

Reply:
Swapped back to the destroy+recreate flavor in the last commit.